[IAP] Check whether transactions in `Transaction.Updates` queue have been handled

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #10187 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
This PR handles In-App Purchases transactions that are currently in the `Transaction.Updates` queue. These transactions could be of several types, for example incomplete purchases, renewals, or switching between plans... and can happen both in-app, or outside the app via the customer App Store. 

When we run the app, we'll be listening to updates to these transactions and handle these accordingly.

Context: pdfdoF-3hA-p2

## Changes:
In the IAP transaction listener, we're now checking for verified transactions, and when we got one in the updates queue, we attempt to reach the `iap/transactions/` endpoint with that transaction ID.

The remote will either return a value (transaction already exists in WPCOM) or throw an unhandled error (transaction does not exist in WPCOM).

- If exists, we just mark the transaction as `.finish()`
- If does not exist, the catch block will be triggered and we'll perform the previously existing `handleCompletedTransaction(:)`

## Testing instructions
Testing this has proven to be tricky, happy to get any feedback on the testing instructions or if there's a better way to do it with Apple built-in StoreKit tools. For now, this seems to work fine:

1. Add a breakpoint in [line 297](https://github.com/woocommerce/woocommerce-ios/blob/1bd40b781bef2f84881fa5939790102488772ce6/Yosemite/Yosemite/Stores/InAppPurchaseStore.swift#L297) and [line 302](https://github.com/woocommerce/woocommerce-ios/blob/1bd40b781bef2f84881fa5939790102488772ce6/Yosemite/Yosemite/Stores/InAppPurchaseStore.swift#L302), so we can stop in the `do` or `catch` blocks depending on the case we're testing. 

2. Using the `WooCommerce IAP` scheme, on a Free Trial site eligible for IAP (a new site can be created in Menu > Switch Sites > Add New Site)
3. In Xcode, open `Debug > StoreKit > Manage Transactions`

<img width="1546" alt="Screenshot 2023-07-19 at 18 51 29" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/7857a7ef-1c53-4950-aa3c-f587d53efec7">

4.  Tap on `Upgrade Now` > select a plan > and see an error (expected since is the testing target and we're using a simulator), and the transaction appears on the Manage Transactions.

<img width="1425" alt="Screenshot 2023-07-19 at 18 52 08" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/30ea8c84-068c-4331-a89f-124e6f7e2120">

5. Re-run the app so the transaction listener is triggered.  Observe that since this is a non existing transaction in the WPCOM billing system, we'll land in the `catch` block, and `handleCompletedTransaction(result)` will be called.

6. In `Debug > StoreKit > Manage Transactions` delete any existing transaction by tapping the "trash" icon, so we can clear the state.

7. In [line 296](https://github.com/woocommerce/woocommerce-ios/blob/1bd40b781bef2f84881fa5939790102488772ce6/Yosemite/Yosemite/Stores/InAppPurchaseStore.swift#L296) change the current `transaction.id` for `UInt64(2000000370069639)`. This is a real transaction that has already been handled in WPCOM billing system:

```diff
	do {
		_ = await self.pauseTransactionListener.values.contains(false)
		// For verified transactions, check whether a transaction has been handled already,
		// and mark it as finished if that's the case, or handle it otherwise
+		let someRealTransactionID = UInt64(2000000370069639)
+		_ = try await self.remote.retrieveHandledTransactionSiteID(for: someRealTransactionID)
-		_ = try await self.remote.retrieveHandledTransactionSiteID(for: transaction.id)
		await transaction.finish()

		self.logInfo("Marking transaction \(transaction.id) as finished")
	}
```
8.  Since transactions are clear, we can upgrade again. Tap on `Upgrade Now` > select a plan > and see an error (expected since is the testing target and we're using a simulator), and a transaction appears on the Manage Transactions.

9. Re-run the app so the transaction listener is triggered.  Observe that since now we have transaction in the queue, but we switched the transaction ID, we'll land in the `do` block, we reach the `retrieveHandledTransactionSiteID` remote and the transaction will be marked as finished since already exists in our end.

10. Note that the console will display the "wrong" transaction ID, since the one printed is from the test transaction we performed through the app, but the one that actually made the flow finish was the one we've injected.
```
2023-07-19 18:52:32.562436+0200 WooCommerce[88904:9172591] [💰IAP Store] Marking transaction 0 as finished
```

iamgabrielma committed:
Update method documentation

iamgabrielma committed:
Rename handling method for clarity

iamgabrielma committed:
Handle transaction verification in StoreKit’s updates queue

toupper commented on Networking/Networking/Remote/InAppPurchasesRemote.swift:
```diff
@@ -90,6 +89,19 @@ public extension InAppPurchasesRemote {
 
+    /// Checks the WPCOM billing system for whether or not an In-app Purchase transaction has been handled
+    ///
+    ///- Returns:The siteID for which the transactionID belongs to
```
```suggestion
    ///- Returns: The siteID the transactionID belongs to
```

toupper commented on Yosemite/Yosemite/Stores/InAppPurchaseStore.swift:
```diff
@@ -283,12 +283,24 @@ private extension InAppPurchaseStore {
+
+                        self.logInfo("Marking transaction \(transaction.id) as finished")
+                    } catch {
+                        _ = await self.pauseTransactionListener.values.contains(false)
```
Do we need to add this step (waiting for the transaction to be finished) twice in the `do` and `catch` blocks?

iamgabrielma replied:
We don't, modified here: https://github.com/woocommerce/woocommerce-ios/pull/10247/commits/edb5c19b292736def74a036ecdfdea040a8c4121

toupper commented on Yosemite/Yosemite/Stores/InAppPurchaseStore.swift:
```diff
@@ -283,12 +283,24 @@ private extension InAppPurchaseStore {
+                        self.logInfo("Marking transaction \(transaction.id) as finished")
+                    } catch {
+                        _ = await self.pauseTransactionListener.values.contains(false)
+                        try await self.handleCompletedTransaction(result)
```
If I understood it correctly, if there's **any** error when getting the handled transaction site id we will try to handle it as completed, thus sending it again through the `orders` endpoint. Is this logic on point? 

E.g. the transaction was handled by WPCOM already, but the `retrieveHandledTransactionSiteID` call failed to a network error. In this case, we will submit it again to orders, which is not the expected action.

Can we check the error and only call `handleCompletedTransaction` if we're sure the transaction wasn't handled in WPCOM?

iamgabrielma replied:
Good point! I modified the remote with https://github.com/woocommerce/woocommerce-ios/pull/10247/commits/05f2e2ce5f5ecfb237b04176fd3bad436219f97f in order to reflect that a successful response might contain two cases (transaction exists, or transaction doesn't exist)  rather than treating the "transaction doesn't exist" case as a side-effect of failing the first case.

Then modified the logic here: https://github.com/woocommerce/woocommerce-ios/pull/10247/commits/edb5c19b292736def74a036ecdfdea040a8c4121 so now we're only calling `handleCompletedTransaction` if the transaction hasn't been handled yet in WPCOM, marking the transaction as finished if could be found, and catching an error if there was any when reaching the remote.

toupper commented on Networking/Networking/Remote/InAppPurchasesRemote.swift:
```diff
@@ -90,6 +89,19 @@ public extension InAppPurchasesRemote {
+    ///
+    ///- Returns:The siteID for which the transactionID belongs to
+    ///
+    @MainActor
```
Just wondering, why do we need this to be run on the Main Actor?

iamgabrielma replied:
We do! Otherwise we'll crash when calling the remote inside `Transactions.Updates` if we happen to have anything in the queue waiting to be processed.

toupper replied:
Oh that's interesting, do you have more information about that crash or documentation from Apple where they talk about using the main thread for this logic?
Even with that, I would decouple the need for execution on the main thread from the function itself, moving it to the function call, something like

`Task { @MainActor in ...
`

That way we can call this function from a background thread if necessary. Even if this is not the case because we only call it there, it's a good practice to decouple and **only** use @MainActor if the function really needs to execute in the main thread, e.g. UI actions.

iamgabrielma replied:
> Oh that's interesting, do you have more information about that crash or documentation from Apple where they talk about using the main thread for this logic?

As far as I can see this bit is not on Apple's end but on ours when dealing with authenticated requests to wpcom:

The crash seems to happens because when we call the remote, there's a call to `WKWebView` that doesn't happen in the main thread, which causes the fatal:

When we create and enqueue the network request as `DotComRequest`, this is constructed by injecting the WPCOM token into an `AuthenticatedDotcomRequest`, which uses the WooCommerce iOS `UserAgent`, which calls a WKWebView instance. And this can only be used from the main thread:

```
-[WKWebView .cxx_construct] must be used from main thread only
```

> Even with that, I would decouple the need for execution on the main thread from the function itself, moving it to the function call

Good call, I moved this to a different method and moved the main thread logic to it. Here: https://github.com/woocommerce/woocommerce-ios/pull/10247/commits/ab3bf022ed87b5ae4a8f6c1cd1a84ccecdb83bda

I think there's room for improvement on the latest changes as well, since we can get the transaction from the result, but that would involve further refactoring when dealing with the different result cases, so I left it outside of this PR for the moment, will open a separate enhancement request for improving that.

iamgabrielma replied:
Tracked follow-up issue here: https://github.com/woocommerce/woocommerce-ios/issues/10294

iamgabrielma committed:
Update IAP remote

With the previous implementation, we could face a decoding or network error but mark it as “unhandled transaction”, which would defeat the purpose of handling the wpcom response for sent transactions. With these changes, we decode both responses (handled vs unhandled) appropiately, and throw an error when needed.

iamgabrielma committed:
Rename method. Add documentation

iamgabrielma committed:
Mark transactions finished if found in wpcom, or handle them

iamgabrielma committed:
Spacing

iamgabrielma committed:
Merge branch 'trunk' into issue/10187-iap-check-if-transaction-is-handled

toupper commented on Networking/Networking/Remote/InAppPurchasesRemote.swift:
```diff
@@ -4,6 +4,7 @@ import Foundation
 /// In-app Purchases Endpoints
 ///
 public class InAppPurchasesRemote: Remote {
+    public typealias InAppPurchasesTransactionResult = Swift.Result<InAppPurchasesTransactionResponse, Error>
```
If the function can throw, we don't need to encapsulate the response in a Result with an error, with can directly make it return the `InAppPurchasesTransactionResponse` instance, or even easier, just the `siteID`

iamgabrielma replied:
I thought about returning only the `siteID`, but since we may have 2 different objects returning a successful response (siteID or message/code) or an Error (decoding/networking/something else) it seemed simpler to return a `Result<Response, Error>` type and hide the details of what that result would be, leaving the business logic to handle it. 

Happy to iterate further if you see a better way to handle this!

toupper replied:
I see, could we then return `InAppPurchasesTransactionResponse` asynchronously from that function? If the function throws an error as well we don't need to return a result as far as I can tell.

iamgabrielma replied:
>  could we then return InAppPurchasesTransactionResponse asynchronously from that function?

Unless I miss-understood, we already do so via the [async wrapper here](https://github.com/woocommerce/woocommerce-ios/blob/ab3bf022ed87b5ae4a8f6c1cd1a84ccecdb83bda/Networking/Networking/Remote/InAppPurchasesRemote.swift#L100C3-L106), since we can't return the original method directly as async as this has to conform to the `Mapper` protocol.

```swift
    func retrieveHandledTransactionResult(for transactionID: UInt64) async throws -> InAppPurchasesTransactionResponse {
        try await withCheckedThrowingContinuation { continuation in
            retrieveHandledTransactionResult(for: transactionID) { result in
                continuation.resume(with: result)
            }
        }
    }
```

Let me know if you're referring to something else! 🙇

toupper replied:
Ah I see, then it's correct, thanks a lot for the answer 👍

iamgabrielma committed:
Extract handling verified transaction result

iamgabrielma committed:
Add internal documentation

toupper commented:
@iamgabrielma The code looks good now, but I don't manage to follow your testing steps. Specifically, when trying a purchase I get a different error and the transaction is not added to the StoreKit Debug Transactions Menu:

<img src="https://github.com/woocommerce/woocommerce-ios/assets/1864060/271ee785-ebf5-42fc-8bde-3735c98c70d4" width="375">

iamgabrielma commented:
> @iamgabrielma The code looks good now, but I don't manage to follow your testing steps. Specifically, when trying a purchase I get a different error and the transaction is not added to the StoreKit Debug Transactions Menu:

Thanks for testing! From the `iap.T.120.2` error message it seems you're falling into the catch block for `inAppPurchaseStoreKitFailed.userCancelled` error. 

- Can you double check you're running the `WooCommerce IAP` scheme? 
- If that's the case, is there any error logged in the console? It should be a "Error purchasing product \(productID) for site \(siteID): \(error)" message
- If all looks as expected, could you invite me to the store so I can take a look?

toupper reviewed (approved):
LGTM and tests well! 🚢 :shipit:
